include/swift/AST/Types.h

```diff
@@ -793,6 +793,22 @@ class alignas(1 << TypeAlignInBits) TypeBase {
   /// `A<Int, NSObject>`.
   Type getSuperclassForDecl(const ClassDecl *classDecl);
 
+  /// \brief Retrieve the closest generic superclass of this type (including the
+  /// type itself).
+  ///
+  /// For example, given
+  ///
+  /// class A<X> {}
+  /// class B<Y> : A<Y> {}
+  /// class C : B<Float> {}
+  ///
+  /// Calling `C`->getLowestGenericClassInHierarchy() returns `B<Float>`, and
+  /// `B<Int>`->getLowestGenericClassInHierarchy() returns `B<Int>`.
+  ///
+  /// \returns The closest superclass of this type that has generic parameters,
+  /// including the type itself, or null if none exists.
+  Type getLowestGenericClassInHierarchy();
```

Comment:
I like the term "closest" that you used in comments better than "lowest", because I've seen people write inheritance in both directions and wouldn't necessarily know which "lower" we're talking about.

Reply:
I ended up changing it to "getGenericAncestor" on @slavapestov's suggestion, since it covers slightly more than generic classes: 

```swift
class Outer<T> { class Inner {} }
class ThisShouldAlsoBeFlagged: Outer<Int>.Inner {}
```